### DISCUSSION TITLE
MySQL to PostgreSQL migration and performance comparison

### DISCUSSION THREAD

**Discussion by james-d-elliott** [authelia/authelia#4523] [General]
In reference to https://github.com/authelia/authelia/issues/4519#issuecomment-1343412655

**Comment by james-d-elliott** [comment]
PostgreSQL has better performance by default when measuring time, and MySQL has better performance by default when measuring memory usage (if I remember the memory usages of each correctly but I will confirm this over the weekend).

As far as a migration path there is no direct one currently, however I'd be willing to make one if users were interested in it (you could just start over, all 2FA's would need registration again however).

**Comment by boomam** [comment]
Morning - whilst migration paths are good, its probably important that we understand what we lose by just creating a new database all-together?   
&nbsp;  

As if its just something as simple as user/s need to re-add their MFA/Duo tokens, then a full guide (Whilst useful to eventually have) becomes something uber-low-priority as the impact is almost zero.   
But if the database is doing something that absolutely cannot be lost as it'll revert Authelia back to 'stock', then a guide should be comparatively higher priority.   
&nsbp;  

Regarding the guide itself, priority aside - a quick search around online shows a few methodologies for migrating between the two, so could certainly be a weekend/I'm bored for 30mins task.  
If I get time this weekend, I'll give it a go myself and add something to the docs. :-)  
(key word here is 'if' :-p)

#EDIT#
Reading around the [docs](https://www.authelia.com/configuration/storage/introduction/), it lists "*user preferences, 2FA device handles and secrets, authentication logs, etc…*" - the question here then becomes how much of that is actually important or not?

**Reply by james-d-elliott** [reply]
In my opinion the important section is 2FA options. So if I were to create export/import methods for those, it'd solve most user issues with migration.

**Comment by james-d-elliott** [comment]
I'm working on creating dedicated export/import cli arguments for this today.

The most essential tables in order of importance:

- user_opaque_identifiers
    - Contains OpenID Connect `sub` values for users which clients often use to link federated users to local users
- webauthn_devices
    - Contains 2FA  Webauthn/Security Key registrations
- totp_configurations
    - Contains 2FA  TOTP registrations

**Reply by boomam** [reply]
Cool, reach out on Discord if you a guinea pig to test. :-)

I was playing with pgloader last night, but it's not ideal. So if you come up with a better way, great!
